Add recipe for defcapture

aggu4 opened the pull request:
### Brief summary of what the package does

Defcapture is a  `defun` analogue and a dedicated namespace for `org-capture` templates. 
It uses the [Doct DSL](https://github.com/progfolio/doct)

### Direct link to the package repository

https://github.com/aggu4/defcapture

### Your association with the package

I am the maintainer.

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

progfolio commented:
Cool idea. I remember when you suggested this a couple years ago to include in DOCT:
https://github.com/progfolio/doct/pull/26

A few notes after a quick review:

I get the following byte-compilation warnings when installing the package:

```emacs-lisp
In defcapture--sync-org-captures:
defcapture.el:2:109 Warning: docstring wider than 80 characters
defcapture.el:9:111 Warning: assignment to free variable ‘org-capture-templates’
```

Unfortunately this doesn't work with all valid DOCT declarations.
e.g.


```emacs-lisp
(progn
  (require 'org-capture)
  (require 'defcapture)
  (let ((org-capture-templates nil)
        (defcapture--capture-namespace (make-hash-table)))
    (defcapture p1 () "First Parent" :keys "1")
    (defcapture c1 (p1) "First Child" :keys "c" :file "")
    (defcapture p2 () "Second Parent" :keys "2")
    (defcapture c2 (p2) "Second Child" :keys "c" :file "")
    org-capture-templates))
```

Results in the following error:

```emacs-lisp
DOCT Declaration has no target: (:clock :file :function :id), nil, ("First Parent" :keys "1")
```

However, a parent declaration needn't have a target if it has children.
You'd have to extend your syntax so that the user can indicate the declaration is a parent before any children are attatched to it.
Likewise, for DOCT's `:group` declarations.
Perhaps the arglist could accept special values to denote these cases. e.g.


```emacs-lisp
(defcapture parent t "Parent" :keys "p") ;; t indicates the user intends to add children.
(defcapture group :group "inherited settings") ;; :group keyword denotes a group.
```

```emacs-lisp
(cl-defstruct defcapture--capture
  parents
  is-child
  declaration)
```

`is-child` is implied by `parents` being non-nil, no?
Checking that slot directly obviates the need for `defcapture--capture-is-child`.


```emacs-lisp
(defun defcapture--capture-boundp (name)
  "Return t if NAME is bound to a capture and nil if it isn't."
  (if (and (gethash name defcapture--capture-namespace) t)
      t
    (progn
      (message "The capture %s isn't defined" name)
      nil)))
```

Can be simplified to:

```emacs-lisp
(defun defcapture--capture-boundp (name)
  "Return t if NAME is bound to a capture and nil if it isn't."
  (or (and (gethash name defcapture--capture-namespace) t)
      (and (message "The capture %s isn't defined" name) nil)))
      ```
      
```emacs-lisp
(defun defcapture--compute-children (name)
(defun defcapture--make-declaration (name)
(defun defcapture-generate-defcapture (name)
```

It's better to name functions based off of what they return, rather than the fact that they "compute", "make", or "generate" something. e.g. 


```emacs-lisp
;; --children-names would be better considering that's what's returned.
(defun defcapture--children (name) ...
(defun defcapture--declaration (name) ...
(defun defcapture-definition (name) ...
```

```emacs-lisp
(defun defcapture--make-declaration (name)
  "Return the declaration for defcapture--capture NAME collecting it's children."
  (let ((children (defcapture--compute-children name)))
    (if children
        (append (defcapture--capture-declaration
                  (gethash name defcapture--capture-namespace))
                (list
                 :children (mapcar #'defcapture--make-declaration
                                   children)))
      (defcapture--capture-declaration
        (gethash name defcapture--capture-namespace)))))
```

Could be simplified to:

```emacs-lisp
(defun defcapture--template (name)
  "Return an Org capture-template for declaration with NAME."
  (if-let ((found (gethash name defcapture--namespace)))
      (append (defcapture--capture-declaration found)
              (list :children (mapcar #'defcapture--template (defcapture--children name))))
    (error "%S not a registered name." name)))
```

I took the liberty of changing the function names in that example.
The term "declaration" is what DOCT considers input for the `doct` function.
The term "capture template" (not to be confused with "template string", which is what gets filled when the template is selected - thank Org for that term being overloaded...) is what the `doct` function outputs.
It should make your code easier to understand by using the same distinction.


```emacs-lisp
(defun defcapture--sync-org-captures ()
  "Sync the captures in `defcapture--capture-namespace' with `org-capture-templates'."
  (setq org-capture-templates
```

I would reconsider setting the global variable by default.
It would be more more flexible to return a list of templates which the user can assign.
The behavior could be changed by a defcustom, which users could opt into if they are sure they want to globally clobber org-capture-templates every time the macro expansion is evaluated.

aggu4 commented:
Glad you like the idea. Thank you for taking the time to review and write notes. I'll work on your suggestions.

riscy commented:
Yes, thanks for this -- and thank you @progfolio, your feedback is always top notch.  I'm seeing the same byte-compile issues on my end.

aggu4 commented:
I updated defcapture and it byte compiles cleanly on my end. 

I addressed the issues pointed out by @progfolio as well by adopting a new strategy for syncing the namespace with the `org-capture-templates` variable. The issue used to be that  in the example:

```
(progn
  (require 'org-capture)
  (require 'defcapture)
  (let ((org-capture-templates nil)
        (defcapture--capture-namespace (make-hash-table)))
    (defcapture p1 () "First Parent" :keys "1")
    (defcapture c1 (p1) "First Child" :keys "c" :file "")
    (defcapture p2 () "Second Parent" :keys "2")
    (defcapture c2 (p2) "Second Child" :keys "c" :file "")
    org-capture-templates))
```

the function `defcapture-sync-captures` was called every time the macro `defcapture` was called, so it would run `doct` before the children were declared. The new strategy involves letting the user call `defcapture-sync-captures` manually after they're finished declaring all children. It will allow the children to be declared without error and, if a declaration really lacks a target, it will let Doct catch the error. The following works for me:

```
(progn
  (require 'org-capture)
  (require 'defcapture)
  (let ((org-capture-templates nil)
        (defcapture--capture-namespace (make-hash-table)))
    (defcapture p1 () "First Parent" :keys "1")
    (defcapture c1 (p1) "First Child" :keys "c" :file "")
    (defcapture p2 () "Second Parent" :keys "2")
    (defcapture c2 (p2) "Second Child" :keys "c" :file "")

    (defcapture-sync-captures)

    org-capture-templates))
```

progfolio commented:
Nice job. From a quick demo, the changes look like an improvement.
A few more notes:

It looks like you may have a stale reference to a previous function name:

```emacs-lisp
defcapture.el:20:92 Warning: the function ‘defcapture--compute-children’ is not known to be defined. 
```

The `:group` DOCT syntax works, which is good:


```emacs-lisp
(progn
  (require 'org-capture)
  (require 'defcapture)
  (let ((org-capture-templates nil)
        (defcapture--capture-namespace (make-hash-table)))
    (defcapture group () :group "hello" :file "/tmp/test.org")
    (defcapture p1 (group) "First Parent" :keys "1")
    (defcapture c1 (p1) "First Child" :keys "c")
    (defcapture p2 (group) "Second Parent" :keys "2")
    (defcapture c2 (p2) "Second Child" :keys "c")
    (defcapture-sync-captures)))
```


When removing a declaration, it's children should be removed recursively.
Otherwise you the user may end up with an unexpected template.
For example:

```emacs-lisp
(progn
  (require 'defcapture)
  (let ((org-capture-templates nil)
        (defcapture--capture-namespace (make-hash-table)))
    ;; :file is meant to be inherited, this should not act
    ;; produce a stand-alone template
    (defcapture p ()   "Parent" :keys "p" :file "test.org")
    (defcapture c (p1) "Child"  :keys "c")
    (defcapture-remove-capture 'c)
    ;; Erroneously returns a template from the parent definition
    (defcapture-sync-captures)))
```

I still think `defcapture-sync-captures` should be replaced by a function which returns the computed capture template list instead of clobbering the global `org-capture-templates` variable.  Users can still assign the return value to `org-capture-templates` easily. e.g.


```emacs-lisp
(setq org-capture-templates (defcapture-templates))
```

Another benefit of that approach is that requiring defcapture won't unnecessarily load org-capture. Note how DOCT does not require org-capture, itself.

riscy commented:
Here's my much less thoughtful feedback of some of the key checklist issues (thanks @progfolio , I think these suggestions are great) --

`defcapture.el` with byte-compile using Emacs 28.1:
```
defcapture.el:92:20: Warning: the function `defcapture--compute-children' is
    not known to be defined.
```

- Add a LICENSE file to the repository
  See: https://github.com/licensee/licensee

aggu4 commented:
Hi, sorry for the delay but I was busy and couldn't address the issues until today.

I updated `defcapture` according to your comments. Now children get removed when using `defcapture-remove-capture` and the strategy to install the captures is:

```emacs-lisp
(setq org-capture-templates (defcapture-captures))
```

riscy commented:
Thank you!  And welcome to MELPA. :)
